### PR TITLE
Improve transfer listener speed

### PR DESCRIPTION
**Pull request by gnodet**
- Try to improve concurrency
- Delegate display to a single thread
- Disable test which is broken with new logic

### CODE REVIEW

**Comment by cstamas** [comment]
+1 but no change detected...
built master with master vs this PR, pretty much same result (build time)

**Review by cstamas — APPROVED** [review]

**Comment by michael-o** [comment]
Can you quatify the speed improvement in percent?

**Comment by cstamas** [comment]
I think this is a good change despite no "speedup" sensed: the original class uses `synchronized` on each listener method, meaning at one time only one thread can enter only one method... and this MAY (most probably is) coggle resolver when uses multiple threads.

**Comment by michael-o** [comment]
> I think this is a good change despite no "speedup" sensed: the original class uses `synchronized` on each listener method, meaning at one time only one thread can enter only one method... and this MAY (most probably is) coggle resolver when uses multiple threads.

Yes, I used `synchronized` back them because it was guaranteed to work properly. I will review this this week.

File: maven-embedder/src/main/java/org/apache/maven/cli/transfer/ConsoleMavenTransferListener.java [hunk 1]
```diff
@@ -34,79 +36,159 @@
...
-                } else {
-                    buffer.append(resourceName, idx + 1, resourceName.length());
+    void update() {
+        synchronized (lock) {
+            try {
+                long t0 = System.currentTimeMillis();
+                while (true) {
+                    lock.wait(100);
+                    long t1 = System.currentTimeMillis();
+                    if (t1 - t0 > MIN_DELAY_BETWEEN_UPDATES) {
```
Comment (michael-o) [inline maven-embedder/src/main/java/org/apache/maven/cli/transfer/ConsoleMavenTransferListener.java:113]: Don't use ? `currentTimeMillis()` for elapsed time. Use `System#nanoTime()`.

Comment (gnodet) [reply maven-embedder/src/main/java/org/apache/maven/cli/transfer/ConsoleMavenTransferListener.java:113]: I really don't need a resolution below 10ms, this is really not time sensitive at all.  I don't have any real problem changing, but what's the reasoning here ?

Comment (michael-o) [reply maven-embedder/src/main/java/org/apache/maven/cli/transfer/ConsoleMavenTransferListener.java]: I am not talking about resolution, but the fact that current millis is wall clock and can run backwards. `nanoTime()` is guaranteed to be at least monotonic.

**Review by cstamas — APPROVED** [review]

**Comment by michael-o** [comment]
> With latest changes (trivial) perf tested locally both, this PR and #1371
> 
> Slight improvement detected:
> 
>     * master: 1:12
> 
>     * both PRs: 1:09
> 
> 
> Then tried some "scale out" but seems throwing threads onto resolver does not help much: https://gist.github.com/cstamas/6b59b26408ec164084b0001685de4ca5
> 
> (note: values above did run tests, so variance is there, and hence, numbers are different. Values from gist are using command from gist).

Don't forget that we buy that we a lot of code complexity. The current code is straight forward.

**Comment by cstamas** [comment]
> Don't forget that we buy that we a lot of code complexity. The current code is straight forward.

Agreed, it is straightforward, but may cause bottleneck (due concurrent invocations, is simple if you sync everything like you did). Hence created alternate proposal https://github.com/apache/maven/pull/1371 that is a plain wrapper that enables writing TransferListeners that runs on single thread. Note: that is just a PoC, but "works for me" :smile:

**Pull request closed without merging**